extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php

```diff
@@ -73,60 +143,99 @@ private function getIso8601Timestamp( $touched ) {
-		$priority = self::PRIORITIES[$ns] ?? self::DEFAULT_PRIORITY;
-
-		$title = Title::newFromText( '$1', $ns );
-		$urlPrefix = str_replace( '$1', '', $title->getFullURL() );
-
-		$dbr = wfGetDB( DB_SLAVE );
-
-		$query = $dbr->select(
-			'page',
-			[
-				'page_title',
-				'page_touched',
-			],
-			[
-				'page_namespace' => $ns,
-				'page_is_redirect' => false,
-			],
-			__METHOD__,
-			[
-				'ORDER BY' => 'page_id',
-				'LIMIT' => 15000,
-			]
-		);
-
-		while ( $row = $dbr->fetchObject( $query ) ) {
-			echo '<url>' . PHP_EOL;
-			echo '<loc>' . $urlPrefix . wfUrlencode( str_replace( ' ', '_', $row->page_title ) ) . '</loc>' . PHP_EOL;
-			echo '<lastmod>' . $this->getIso8601Timestamp( $row->page_touched ) . '</lastmod>' . PHP_EOL;
-			echo '<priority>' . $priority . '</priority>' . PHP_EOL;
-			echo '</url>' . PHP_EOL;
+	private function generateSitemapPage( array $nses, $page ) {
+		$out = '<?xml version="1.0" encoding="UTF-8"?>' . PHP_EOL;
+		$out .= '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">' . PHP_EOL;
+		$out .= '<!-- Namespaces: ' . implode( ', ', $nses ) . '; page: ' . $page . ' -->' . PHP_EOL;
+
+		$offset = self::URLS_PER_PAGE * ( $page - 1 );
+		$limit = self::URLS_PER_PAGE;
+
+		foreach ( $nses as $ns ) {
+			$priority[$ns] = self::PRIORITIES[$ns] ?? self::DEFAULT_PRIORITY;
+			$title = Title::newFromText( '$1', $ns );
+			$urlPrefix[$ns] = str_replace( '$1', '', $title->getFullURL() );
 		}
+
+		foreach ( $this->model->getItems( $nses, $offset, $limit ) as $page ) {
+			$encodedTitle = wfUrlencode( str_replace( ' ', '_', $page->page_title ) );
+			$lastmod = $this->getIso8601Timestamp( $page->page_touched );
+
+			$out .= '<url>' . PHP_EOL;
+			$out .= '<loc>' . $urlPrefix[$page->page_namespace] . $encodedTitle . '</loc>' . PHP_EOL;
+			$out .= '<lastmod>' . $lastmod . '</lastmod>' . PHP_EOL;
+			$out .= '<priority>' . $priority[$page->page_namespace] . '</priority>' . PHP_EOL;
+			$out .= '</url>' . PHP_EOL;
+		}
+
+		$out .= '</urlset>' . PHP_EOL;
+		return $out;
+	}
+
+	private function generateSitemapIndex() {
+		global $wgServer;
+
+		$out = '<?xml version="1.0" encoding="UTF-8"?>' . PHP_EOL;
+		$out .= '<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">' . PHP_EOL;
+
+		$baseUrl = $wgServer;
+
+		foreach ( self::SEPARATE_SITEMAPS as $ns ) {
+			$ns_pages = $this->model->getPageNumber( [ $ns ], self::URLS_PER_PAGE );
+			for ( $page = 1; $page <= $ns_pages; $page++ ) {
+				$url = $baseUrl . '/sitemap-newsitemapxml-NS_' . $ns . '-p' . $page . '.xml';
+				$out .= '<sitemap><loc>' . $url . '</loc></sitemap>' . PHP_EOL;
+			}
+		}
+
+		$other_pages = $this->model->getPageNumber( $this->getOtherNamespaces(), self::URLS_PER_PAGE );
+
+		for ( $page = 1; $page <= $other_pages; $page++ ) {
+			$url = $baseUrl . '/sitemap-newsitemapxml-other-p' . $page . '.xml';
+			$out .= '<sitemap><loc>' . $url . '</loc></sitemap>' . PHP_EOL;
+		}
+
+		$out .= '</sitemapindex>' . PHP_EOL;
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "$out". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1488)